File: src/main/python/binary_search/my_sqrt.py
class MySqrt:
    def mySqrt(self, x: int) -> int:
        left, right = 0, x
        while left <= right:
            mid = left + (right - left) // 2
            if mid * mid == x: return mid
            if x > mid * mid: left = mid + 1
            else: right = mid - 1
        return left - 1

if __name__ == '__main__':
    s = MySqrt()
    e1 = s.mySqrt(4)
    print(e1 == 2)
    e2 = s.mySqrt(8)
    print(e2 == 2, e2)
    e3 = s.mySqrt(9)
    print(e3 == 3, e3)
    e4 = s.mySqrt(10)
    print(e4 == 3, e4)

File: src/main/java/org/parthvnp/BinarySearch/MySqrt.java
package org.parthvnp.BinarySearch;

public class MySqrt {
  public int mySqrt(int x) {
    long left = 0, right = x;
    while (left <= right) {
      long mid = left + (right - left) / 2;
      if (mid * mid == x) return (int) mid;
      if (x > mid * mid) left = mid + 1;
      else right = mid - 1;
    }
    return (int) left - 1;
  }

  public static void main(String[] args) {
    var s = new MySqrt();
    var e1 = s.mySqrt(2147395599);
    System.out.println(e1);
  }
}
